### PR TITLE
[WIP] Parse yggdrasil models

### PR DESCRIPTION
**Pull request by Xyzic**
Still working on getting other models cleaned up. Only works for photosynthesis.yaml right now

### CODE REVIEW

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:44]: So I *think* that with the new `TraitType` implementation, you should be able to avoid a large part of this.  You should be able to supply a dictionary (there's an example notebook showing this, I believe) to the input slot attribute on a node or component, and it will construct it appropriately.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:50]: I think this is simple enough it could just be a set of `.replace()` calls, right?  Does it need to be a regex?

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:59]: So, even if you weren't able to supply a dictionary, I do think calling the function is unnecessary here.  You already have what amounts to a set of conditionals, and you could just create the node element directly here and not supply the `"i"` and `"o"` (below).

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:66]: I believe this will also take a dictionary and auto-convert it, but even if not, a temporary variable would be fine here.  Do we need to supply the sockets?  I thought the editor tracked those.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:68]: This could be an f-string or a formatted string, but what result does it provide?

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:13]:
```suggestion
with open(ex_yamls['fakeplant']['python'], "r") as test_model:
    photosynthesis_model = yaml.safe_load(test_model)['model']
```

We don't need to declare it since we're redefining it inside the `with`

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:34]: I think this should be in a function:

```suggestion
def load_example(ps = None):
    ps = ps or jlne.NodeEditor()
    dict_conversion(photosynthesis_model)
    ps.add_component(photosynthesis_model)
```
I'm not super keen on the in-place modification, though -- see my comments above.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:16]: I think I would like it if we changed this to return a new version.  That would mean that we accept `model_dict` as input, and then everywhere you've got a `.pop()` that doesn't get assigned, you could skip that line.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:20]: Could we use a variable other than `i`?  I usually think of `i` as an idiom for an iteration counter, but here it's the actual object.  Maybe here it could be `inp` and below it could be `out`.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (Xyzic) [reply jupyterlab_nodeeditor/yggdrasil_support.py:16]: Fixed everything else locally, but just to clarify on this, you mean return a new dict right? Instead of modifying the input and that way we can just pull and copy data instead of deleting existing

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:35]:
```suggestion
    for i, inp in enumerate(model_dict["inputs"]):
```

Comment (matthewturk) [reply jupyterlab_nodeeditor/yggdrasil_support.py:35]: and then thread this through

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:17]: I don't think we need this, do we?  It's initializing a set of dictionaries and empties and whatnot, but can't we just do this inline in the function where we do the iteration?

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:11]: Put this in the function down below -- otherwise it happens at import time

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:43]: same thing here with enumerate

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:36]: if you get rid of the pre-initializer, you can move `new_dict["inputs"] = []` above the loop, and then in here `.append( {"title": ... , "key" ... } )` etc.

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:40]: `title` shows up twice?

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:29]: don't need these

**Comment by Xyzic** [comment]
Updated a lot for review

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:1]: If these are still valid and necessary for the PR to be accepted, can you make them into todo items in the PR body?

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:32]: Why are these getters necessary?

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:51]: Why is this necessary?

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:54]:
```suggestion
    # Test to make sure there are inputs, outputs, and names
```

we should match indent here

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:57]: Let's just move the model_keys inline to the `for` statement line

File: jupyterlab_nodeeditor/backup_framework.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/backup_framework.py:72]: In general this should be a command line flag, not an input.  But that might be out of scope here.  What I would instead suggest is to split this main function and then based on the result of input, call one of the two.  That would help move it toward a more modular approach when we add command line arguments.

File: jupyterlab_nodeeditor/test_sample.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/test_sample.py:8]: We should get rid of the junk functions before this is merged

**Review by matthewturk — COMMENTED** [review]
getting there!

**Comment by Xyzic** [comment]
Hey Matt, I moreso need you to go over the ygg_support.ipnyb as it has most, if not all, of the major changes.
I will make a separate file for the testing functions without using classes for the time being until I can really work through how to integrate pytest into this since I don't have many tests prepared in my mind

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:15]: use format strings or f-strings

```suggestion
        new_dict["inputs"].append({'title': inp["name"], 'key': f"temp_in{i}", 'socket_type': inp["default_file"]["filetype"]})
```

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:19]:
```suggestion
        new_dict["outputs"].append({'title': out["name"], 'key': f"temp_out{o}", 'socket_type': out["default_file"]["filetype"]})
```

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (matthewturk) [inline jupyterlab_nodeeditor/yggdrasil_support.py:45]: I'm not sure this'll ever pass.  But, you might be able to compare the configs against each other.

**Review by matthewturk — APPROVED** [review]
almost there

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (Xyzic) [reply jupyterlab_nodeeditor/yggdrasil_support.py:45]: Oh yeah, I missed this on cleanup, that was supposed to be a palceholder until I had the configs calling nailed down. The test should never pass until I fixed it, will either comment it out or remove it before we merge based on what you think

**Comment by Xyzic** [comment]
Deleted the script file for now, the notebook is here. Once it's all cleared I will convert it to a script instead of an ipynb

File: jupyterlab_nodeeditor/yggdrasil_support.py
Comment (Xyzic) [reply jupyterlab_nodeeditor/yggdrasil_support.py]: Never posted it as a question so allow me to clarify: Would you like it commented or removed?

Comment (matthewturk) [reply jupyterlab_nodeeditor/yggdrasil_support.py]: If it shows up in the test runner then we shouldn't merge it -- comment the line out and assert True below it.

Comment (Xyzic) [reply jupyterlab_nodeeditor/yggdrasil_support.py]: I removed the entire file because it was supposed to not be in here, so it shouldn't be merged at all now

**Pull request merged**
